#159204718 user reset password upon request

yomigeek opened the pull request:
#### What does this PR do?
user should be able to reset their password
send reset link to user email on request
verify reset mail token
send confirmation mail on password reset update

#### Description of Task to be completed?
add sendMail helper method to send emails with sendGrid API
add sendResetEmail, verifyPasswordResetToken, and updateUserPassword methods in Password reset controller add an emailMessages helper method to store all email templates write unit tests to test the functionalities.

#### How should this be manually tested?
upon password reset request, an email containing link to the reset password will be sent to the user email address , when clicked, the token generated is been verified. After which the user are redirected to link containing a token for them to reset their password.

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/159204718

yomigeek committed:
feat(passwordReset): implement user password reset
- implement unit test
- implement validatorjs package to handle user validations
- add PasswordResetController.js file

yomigeek committed:
feat(passwordReset): implement user password reset
- implement unit test
- implement validatorjs package to handle user validations
- add PasswordResetController.js file

yomigeek committed:
feat(passwordReset): implement user password reset
- implement further unit tests
- add send email method to the passwordResetController.js file to send user reset mail with a jwt token

yomigeek committed:
feat(passwordReset): implement user password reset
- implement further unit tests
- add send email method to the passwordResetController.js file to send user reset mail with a jwt token

yomigeek committed:
feat(passwordReset): implement user password reset
- implement further unit tests on verifying reset password route
- add verifyResetPassword method to the passwordResetController.js file to verify the link on user click

yomigeek committed:
feat(passwordReset): implement user password reset
- implement further unit tests on verifying reset password route
- add verifyResetPassword method to the passwordResetController.js file to verify the link on user click

yomigeek committed:
feat(passwordReset): implement user password reset
- rebase the branch from recent develop

[Delivers #159204718]

yomigeek committed:
feat(passwordReset): implement user password reset
- rebase the branch from recent develop

[Delivers #159204718]

yomigeek committed:
feat(passwordReset): implement user password reset

- fix travis build fail

[Delivers #159204718]

yomigeek committed:
feat(passwordReset): implement user password reset

- fix travis build fail

[Delivers #159204718]

yomigeek committed:
Merge branch 'develop' into feature/159204718/user-reset-password

yomigeek committed:
feat(passwordReset): fix travis build fail

 - include missing validation method and fix travis build fail

[Delivers #159204718]

yomigeek committed:
feat(passwordReset): fix travis build fail

 - include missing validation method and fix travis build fail

[Delivers #159204718]

yomigeek committed:
feat(passwordReset): implement user password reset
 - fix travis build fail

[Delivers #159204718]

yomigeek committed:
feat(passwordReset): implement travis build fail

 - fix travis build fail

[Delivers #159204718]

yomigeek committed:
feat(passwordReset): implement travis build fail

 - fix travis build fail

[Delivers #159204718]

yomigeek committed:
feat(passwordReset): implement changes on password reset codebase

- refactor the passwordResetController.js file
- edit unit test
- update Readme file

[Delivers #159204718]

yomigeek committed:
feat(passwordReset): implement changes on password reset codebase

- refactor the passwordResetController.js file
- edit unit test
- update Readme file

[Delivers #159204718]

yomigeek committed:
Merge branch 'develop' into feature/159204718/user-reset-password

madeofhuman commented on package.json:
```diff
@@ -48,6 +51,7 @@
     "pg-hstore": "^2.3.2",
     "request": "^2.87.0",
     "sequelize": "^4.38.0",
+    "should": "^13.2.3",
```
I understand why you felt you needed to install this package, but we're already using Chai, which has a `should` function. That makes this package redundant.

Oluwafayokemi reviewed (commented):
Good job Yomi!
Can you please look into some of the description messages? I really think they can be more descriptive for the user to understand. Also, I see you added the seeder to the migrate:test script. I think it would be better if you created another script and call it seed:test. Weldone again. Great Job!

madeofhuman commented on server/controllers/PasswordResetController.js:
```diff
@@ -0,0 +1,122 @@
+          message: 'Email could not be sent. Please try again',
+        });
+      })
+      .catch(err => res.status(500).send({ errors: { message: err.message } }));
```
Nice error handling! However, there's already an error handler, so you only need to do `.catch(err => next(err);`

madeofhuman reviewed (commented):
Nice work, @yomigeek! Please find time to address the reviews I left.

Veraclins commented on server/controllers/PasswordResetController.js:
```diff
@@ -0,0 +1,122 @@
+          return res.status(201).json({
+            status: 'success',
+            message,
+            token
```
Hi @yomigeek good job with this PR. I noticed that you returned the token you are send ing in the email to the user. This is a security issue as you are effectively giving him the key without him needing to go into his email. I think you should remove it.

chukwuemekachm reviewed (changes requested):
Nice job on completing this task. I see that all the feedback has already been given. However, you should try to fix all the issues with Hound CI. You might also want to add a new line at the end of `.env.example` file to avoid red squigglies. Well done.

yomigeek committed:
feat(passwordReset): implement feedback on password reset codebase

- make changes to responses to the users on the passwordResetController.js file
- refactor unit tests

[Delivers #159204718]

yomigeek committed:
feat(passwordReset): implement feedback on password reset codebase

- make changes to responses to the users on the passwordResetController.js file
- refactor unit tests

[Delivers #159204718]

yomigeek committed:
feat(passwordReset): implement changes on password reset codebase
- update local branch from remote branch

[Delivers #159204718]

yomigeek committed:
feat(passwordReset): implement changes on password reset codebase

- update changes to password reset  branch
- fix travis build issues

[Delivers #159204718]

yomigeek committed:
feat(passwordReset): implement changes on password reset codebase

- update changes to password reset  branch
- fix travis build issues

[Delivers #159204718]

yomigeek committed:
feat(passwordReset): implement changes on password reset codebase

- refactor the codebase for travis to build successfully
- edit unit test

[Delivers #159204718]

yomigeek committed:
feat(passwordReset): implement changes on password reset codebase

- refactor the codebase for travis to build successfully
- edit unit test

[Delivers #159204718]

Oluwafayokemi committed:
#159204719 Enable social login (#11)

* feat(social-login): enable social login
- implement passport authentication for google
- implement passport authentication for facebo:ok
- implement passport authentication for twitter
- unit test for social login authentication
[Delivers #159204719]

* validating social authentication route for google and facebook

* return token with authenticated user response

* implement a mock test strategy for the authenticatio

* feat(social-login): enable social login
- implement passport authentication for google
- implement passport authentication for facebo:ok
- implement passport authentication for twitter
- unit test for social login authentication
[Delivers #159204719]

* validating social authentication route for google and facebook

* return token with authenticated user response

* implement a mock test strategy for the authenticatio

* rebase file and fix merge conflict

* feat(social-login): enable social login
- implement passport authentication for google
- implement passport authentication for facebo:ok
- implement passport authentication for twitter
- unit test for social login authentication
[Delivers #159204719]

* validating social authentication route for google and facebook

* return token with authenticated user response

* implement a mock test strategy for the authenticatio

* feat(social-login): enable social login
- implement passport authentication for google
- implement passport authentication for facebo:ok
- implement passport authentication for twitter
- unit test for social login authentication
[Delivers #159204719]

* validating social authentication route for google and facebook

* return token with authenticated user response

* implement a mock test strategy for the authenticatio

* rebase file and fix merge conflict

* rebase on password verification

* change validateController file name to verfyController file name

* - remvoed commented codes -wrote descriptive message errors

* remove unknown instabul plugin fron.babelrc

* resolve verifyController files
write simple test for invalid facebook callback token

* remove a few extra comments and update .env_example file

* fix email verificication issue

* --amend

* fix hound issues

yomigeek committed:
feat(passwordReset): implement changes on password reset codebase

- refactor the codebase for travis to build successfully
- edit unit test

[Delivers #159204718]

yomigeek committed:
feat(passwordReset): implement changes on password reset codebase

- refactor the codebase for travis to build successfully
- edit unit test

[Delivers #159204718]

yomigeek committed:
feat(passwordReset): rebase changes into password reset branch

- rebase and refactor the codebase for travis to build successfully

[Delivers #159204718]

yomigeek committed:
feat(passwordReset): resolve merge conflict issues

- resolve merge conflict from remote branch to local branch

yomigeek committed:
feat(passwordReset): resolve merge conflict issues

- resolve merge conflict from remote branch to local branch

yomigeek committed:
feat(passwordReset): resolve merge conflict issues

- resolve merge conflict

[Delivers #159204718]

Oluwafayokemi reviewed (approved):
Good Job Yomi!

Veraclins commented on .babelrc:
```diff
@@ -4,9 +4,6 @@
     "retainLines": true,
     "env": {
         "test": {
-          "plugins": [
```
This plugin makes it possible for coverage reports to be displayed. So removing it means the report will not be generated. Please add it back

yomigeek replied:
Thanks for that Clinton. Will add it back.

chukwuemekachm commented on server/controllers/PasswordResetController.js:
```diff
@@ -0,0 +1,119 @@
+        email,
+      }
+    }).then((updatePassword) => {
+      if (updatePassword) {
```
Nice job on completing this task. However, 
- On this line, You are checking if the user exists before updating the database. But no code handles when the user doesn't exist. I feel this might stall the request on that middleware without returning a response.

yomigeek replied:
Thanks, will look into that. I appreciate the review!

chukwuemekachm commented on server/migrations/20180716204138-create-user.js:
```diff
@@ -0,0 +1,65 @@
+      type: Sequelize.BOOLEAN,
+      defaultValue: false,
+    },
+    bio: {
```
Please the user's bio datatype should be a `TEXT` and not a `STRING`, this is because a `STRING` is limited to about 255 characters while `TEXT` can store up to 5,000 characters.

yomigeek replied:
Thanks for the review, however, if you check the code base for the user reset on the branch you will notice it has been updated to TEXT after rebasing process. Thanks.

Veraclins commented on server/controllers/PasswordResetController.js:
```diff
@@ -0,0 +1,119 @@
+    return res.status(200).send({
+      status: 'success',
+      message: 'Please use the link below to change your password',
+      token: passwordUpdateToken,
```
@yomigeek, nice work you are doing. I feel that returning the URL is enough and no need to return the token as well. You may consider removing it as it exposes too much about the internal process.

Veraclins reviewed (commented):
Good job @yomigeek. You really did a nice job here. Just fix the few issues pointed out and I believe you are good to go.

madeofhuman commented on server/config/database.js:
```diff
@@ -1,4 +1,6 @@
-require('dotenv/config');
+const dotenv = require('dotenv');
+
+dotenv.config();
```
@yomigeek you've done a nice work in completing this task. 🙌🏾
I would like to know, though, your reasoning behind this import modification because I feel you basically did the same thing in a longer way.

madeofhuman reviewed (commented):
Good job completing the task, Yomi! 🙌🏾
I noticed you used the es5 method of module importation in some places, especially with `dotenv` imports. You could try standardizing them with the es6 methods used in the rest of the code, for uniformity.
You can try this `import { } from 'dotenv/config';` for `dotenv` imports.

yomigeek committed:
feat(passwordReset): implement feedback on password reset codebase

- make changes to the passwordResetController.js file based on peer reviews
- refactor unit tests

[Delivers #159204718]

chukwuemekachm commented:
@yomigeek  You have done a great job by fixing all the Hound CI violations and implementing the feedback provided by the team. I noticed that there is a `10%` code coverage decrease on your feature branch, do you mind explaining what happened?

Veraclins commented:
Good job @yomigeek. @chukwuemekachm I don't think the reduction is really from him, It is showing now simply because the coverage was not building at runtime because the istanbul plugin was removed in the previous PR. This resulted in a false positive previously. Hence the decrease now that it is building and reflecting the true coverage.

tonymontaro commented on server/controllers/PasswordResetController.js:
```diff
@@ -0,0 +1,127 @@
+    // Jwt token
+    const token = JwtHelper.createToken({ email }, 1800);
+    // Create Reset Email Data
+    // This is to verify if user email exist and send reset
```
Excellent work @yomigeek! Please remove these unnecessary comments. The JSdoc should contain enough information about the class/function. Also, the code is already (or should be) self-explanatory.

yomigeek committed:
feat(passwordReset): implement feedback on password reset codebase
- make changes to the codebase files based on peer reviews
- remove unnecessary comments

[Delivers #159204718]
